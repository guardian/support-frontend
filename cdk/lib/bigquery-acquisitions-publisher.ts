import { GuAlarm } from "@guardian/cdk/lib/constructs/cloudwatch";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuLambdaFunction } from "@guardian/cdk/lib/constructs/lambda";
import type {App} from "aws-cdk-lib";
import { aws_events, aws_sqs, Duration } from "aws-cdk-lib";
import {
  ComparisonOperator,
  TreatMissingData,
} from "aws-cdk-lib/aws-cloudwatch";
import { Archive, EventBus, Rule } from "aws-cdk-lib/aws-events";
import { SqsQueue } from "aws-cdk-lib/aws-events-targets";
import { PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { LoggingFormat, Runtime } from "aws-cdk-lib/aws-lambda";
import { SqsEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import { Queue } from "aws-cdk-lib/aws-sqs";

export interface BigqueryAcquisitionsPublisherProps extends GuStackProps {
  stack: string;
  stage: string;
  softOptInConsentSetterQueueArn: string;
}

const appName = "bigquery-acquisitions-publisher";

export class BigqueryAcquisitionsPublisher extends GuStack {
  constructor(scope: App, id: string, props: BigqueryAcquisitionsPublisherProps) {
    super(scope, id, props);

    const busName = `acquisitions-bus-${props.stage}`;

    // Event bus
    const eventBus = new EventBus(this, busName, {
      eventBusName: busName,
    });

    new Archive(this, `${busName}-archive`, {
      eventPattern: {
        account: ["account"],
        region: ["eu-west-1"],
      },
      sourceEventBus: eventBus,
      archiveName: `${busName}-archive`,
      description: `Archive for all events sent to ${busName}-archive`,
      retention: Duration.days(90),
    });

    // SQS Queues
    const queueName = `${appName}-queue-${props.stage}`;
    const deadLetterQueueName = `dead-letters-${appName}-${props.stage}`;

    const deadLetterQueue = new Queue(this, `dead-letters-${appName}Queue`, {
      queueName: deadLetterQueueName,
      retentionPeriod: Duration.days(14),
    });

    const acquisitionsQueue = new Queue(this, `${appName}Queue`, {
      queueName,
      visibilityTimeout: Duration.minutes(2),
      deadLetterQueue: {
        // The number of times a message can be unsuccessfully dequeued before being moved to the dead-letter queue.
        // This has been set to 1 to avoid duplicate acquisition events being sent to bigquery
        maxReceiveCount: 1,
        queue: deadLetterQueue,
      },
    });

    // Rule which passes events on to SQS
    new Rule(this, "EventBusToSQSRule", {
      description: "Send all events to SQS",
      eventPattern: {
        region: ["eu-west-1"],
      },
      eventBus: eventBus,
      targets: [new SqsQueue(acquisitionsQueue)],
    });

    const softOptInConsentSetterQueue =
      aws_sqs.Queue.fromQueueArn(this, "SoftOptInConsentSetterQueue", props.softOptInConsentSetterQueueArn)

    // Rule which passes events from support-workers on to soft opt-in queue
    new Rule(this, "SoftOptInToSQSRule", {
      description: "Send all events received via support-workers onto soft opt-in SQS queue",
      eventBus: eventBus,
      eventPattern: {
        region: ["eu-west-1"],
        source: ["support-workers.1"],
      },
      targets: [
        new SqsQueue(
          softOptInConsentSetterQueue,
          {
            message: aws_events.RuleTargetInput.fromObject({
              subscriptionId: aws_events.EventField.fromPath('$.detail.zuoraSubscriptionNumber'),
              identityId: aws_events.EventField.fromPath('$.detail.identityId'),
              eventType: "Acquisition",
              productName: aws_events.EventField.fromPath('$.detail.ZuoraProductName'),
              previousProductName: null
            }),
          },
        )
      ],
    });

    // Create a custom role because the name needs to be short, otherwise the request to Google Cloud fails
    const role = new Role(this, "bigquery-to-s3-role", {
      roleName: `bq-acq-${this.stage}`,
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
    });
    role.addToPolicy(
      new PolicyStatement({
        actions: [
          "logs:CreateLogGroup",
          "logs:CreateLogStream",
          "logs:PutLogEvents",
        ],
        resources: ["*"],
      })
    );
    role.addToPolicy(
      new PolicyStatement({
        actions: ["ssm:GetParameter"],
        resources: [
          `arn:aws:ssm:${this.region}:${this.account}:parameter/${appName}/${props.stage}/gcp-wif-credentials-config`,
        ],
      })
    );

    const monitoring =
      this.stage == "PROD"
        ? {
            toleratedErrorPercentage: 0,
            snsTopicName: `alarms-handler-topic-${this.stage}`,
            alarmName: "big-query-acquisition-publisher lambda has failed",
            alarmDescription:
              "Check the logs for details https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Fbigquery-acquisitions-publisher-PROD",
          }
        : undefined;

    // SQS to Lambda event source mapping
    const eventSource = new SqsEventSource(acquisitionsQueue, {
      reportBatchItemFailures: true,
    });

    const functionName = `${appName}-${props.stage}`;

    new GuLambdaFunction(this, `${appName}Lambda`, {
      app: appName,
      runtime: Runtime.NODEJS_20_X,
      fileName: `index.zip`,
      functionName,
      handler: "index.handler",
      events: [eventSource],
      timeout: Duration.minutes(2),
      role,
      errorPercentageMonitoring: monitoring,
      loggingFormat: LoggingFormat.TEXT,
    });

    new GuAlarm(this, "DeadLetterQueueAlarm", {
      app: appName,
      alarmName: `The ${props.stage} ${appName} lambda has failed`,
      alarmDescription:
        `There is one or more event in the ${deadLetterQueueName} dead letter queue (DLQ). ` +
        "Check the logs for the error and use the details to confirm that the event was not written " +
        "to the fact_acquisition_event table in BigQuery. If the event is not in the table then use the DLQ " +
        "redrive feature to replay the failed event. If the redrive functionality is not used " +
        "then purge the queue instead or the alarm will remain in an alarm state.\n" +
        `Main queue: https://eu-west-1.console.aws.amazon.com/sqs/v2/home?region=eu-west-1#/queues/https%3A%2F%2Fsqs.eu-west-1.amazonaws.com%2F865473395570%2F${queueName}\n` +
        `DLQ: https://eu-west-1.console.aws.amazon.com/sqs/v2/home?region=eu-west-1#/queues/https%3A%2F%2Fsqs.eu-west-1.amazonaws.com%2F865473395570%2F${deadLetterQueueName}\n` +
        `Logs: https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252F${functionName}`,
      metric: deadLetterQueue.metricApproximateNumberOfMessagesVisible(),
      threshold: 1,
      evaluationPeriods: 1,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: TreatMissingData.IGNORE,
      snsTopicName: `alarms-handler-topic-${this.stage}`,
      actionsEnabled: this.stage === "PROD",
    });
  }
}
