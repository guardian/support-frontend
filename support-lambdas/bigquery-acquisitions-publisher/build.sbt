import LibraryVersions.circeVersion
import LibraryVersions._

name := "bigquery-acquisitions-publisher"
description := "A lambda for publishing acquisitions events to BigQuery"

libraryDependencies ++= Seq(
  "com.amazonaws" % "aws-lambda-java-core" % "1.2.0",
  "com.amazonaws" % "aws-lambda-java-events" % "3.11.1",
  "com.amazonaws" % "aws-java-sdk-ssm" % awsClientVersion,
  "ch.qos.logback" % "logback-classic" % "1.5.6",
  "com.typesafe.scala-logging" %% "scala-logging" % "3.9.5",
  "io.circe" %% "circe-core" % circeVersion,
  "io.circe" %% "circe-generic" % circeVersion,
)

assemblyJarName := s"${name.value}.jar"
riffRaffPackageType := assembly.value
riffRaffUploadArtifactBucket := Option("riffraff-artifact")
riffRaffUploadManifestBucket := Option("riffraff-builds")
riffRaffManifestProjectName := s"support:lambdas:${name.value}"
riffRaffArtifactResources += (file(
  "cdk/cdk.out/BigqueryAcquisitionsPublisher-PROD.template.json",
), "cfn/BigqueryAcquisitionsPublisher-PROD.template.json")
riffRaffArtifactResources += (file(
  "cdk/cdk.out/BigqueryAcquisitionsPublisher-CODE.template.json",
), "cfn/BigqueryAcquisitionsPublisher-CODE.template.json")

lazy val deployToCode =
  inputKey[Unit]("Directly update AWS lambda code from local instead of via RiffRaff for faster feedback loop")

deployToCode := {
  import scala.sys.process._
  val s3Bucket = "membership-dist"
  val s3Path = "support/CODE/bigquery-acquisitions-publisher/bigquery-acquisitions-publisher.jar"
  (s"aws s3 cp ${assembly.value} s3://" + s3Bucket + "/" + s3Path + " --profile membership --region eu-west-1").!!
  List(
    "bigquery-acquisitions-publisher-CODE",
  ).foreach(functionPartial => {
    System.out.println(s"Updating function $functionPartial")
    s"aws lambda update-function-code --function-name $functionPartial --s3-bucket $s3Bucket --s3-key $s3Path --profile membership --region eu-west-1".!!
  })
}
