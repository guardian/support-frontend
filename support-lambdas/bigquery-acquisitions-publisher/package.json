{
	"name": "bigquery-acquisitions-publisher",
	"version": "1.0.0",
	"main": "index.js",
	"license": "MIT",
	"scripts": {
		"tsc": "tsc --noEmit",
		"test": "vitest --run",
		"test-watch": "vitest",
		"build": "esbuild --bundle --platform=node --target=node20 --outfile=target/typescript/index.js typescript/index.ts"
	},
	"devDependencies": {
		"@guardian/prettier": "^8.0.1",
		"@types/aws-lambda": "^8.10.147",
		"@types/node": "^22.13.4",
		"esbuild": "^0.25.0",
		"prettier": "^3.5.1",
		"tslib": "^2.6.2",
		"typescript": "^5.7.3",
		"vitest": "^3.0.7"
	},
	"dependencies": {
		"@aws-sdk/client-ssm": "^3.893.0",
		"@aws-sdk/credential-provider-node": "^3.750.0",
		"google-auth-library": "^9.0.0",
		"@google-cloud/bigquery": "^7.9.2",
		"aws-lambda": "^1.0.7",
		"zod": "^3.24.2"
	}
}
