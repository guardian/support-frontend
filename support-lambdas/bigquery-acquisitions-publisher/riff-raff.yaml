stacks: [support]
regions: [eu-west-1]

deployments:
  bigquery-acquisitions-publisher:
    type: aws-lambda
    parameters:
      bucket: membership-dist
      functionNames: [  bigquery-acquisitions-publisher-]
      fileName: index.zip
      prefixStack: false
    dependencies: [ cfn ]
  cfn:
    type: cloud-formation
    app: bigquery-acquisitions-publisher
    parameters:
      templateStagePaths:
        CODE: BigqueryAcquisitionsPublisher-CODE.template.json
        PROD: BigqueryAcquisitionsPublisher-PROD.template.json
      prependStackToCloudFormationStackName: false
