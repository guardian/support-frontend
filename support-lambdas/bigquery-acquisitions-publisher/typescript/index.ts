import type { SQSEvent } from 'aws-lambda';
import { transformAcquisitionProductForBigQuery } from './acquisitions';
import {
	buildAuthClient,
	createBigQueryClient,
	writeRowToBigQuery,
} from './bigQuery';
import { AcquisitionProductEventSchema } from './schemas';
import { getGCPCredentialsFromSSM } from './ssm';
import { getStage } from './stage';

type Result =
	| {
			success: true;
	  }
	| {
			success: false;
			messageId: string;
	  };

export const handler = async (event: SQSEvent) => {
	// BigQuery connect
	const stage = getStage();
	const credentials = await getGCPCredentialsFromSSM(stage);
	const authClient = await buildAuthClient(credentials);
	const bigQueryClient = createBigQueryClient(authClient, stage);

	// Parse SQSEvents
	console.log('Received event:', event);
	const asyncResults: Array<Promise<Result>> = event.Records.map(
		async (record): Promise<Result> => {
			// JSON parse record body
			let payload;
			try {
				payload = JSON.parse(record.body) as unknown;
			} catch (jsonError) {
				if (jsonError instanceof Error) {
					console.error('Failed to JSON parse payload:', jsonError.message);
				}
				return { success: false, messageId: record.messageId };
			}

			// Validate the payload against Zod schema
			const parsedAcquisitionProductDetail =
				AcquisitionProductEventSchema.safeParse(payload);
			if (!parsedAcquisitionProductDetail.success) {
				console.error(
					'Failed to Schema parse payload:',
					parsedAcquisitionProductDetail.error,
				);
				return { success: false, messageId: record.messageId };
			}

			// Transform payload to BigQuery row
			const row = transformAcquisitionProductForBigQuery(
				parsedAcquisitionProductDetail.data.detail,
			);

			// Write BigQuery row
			try {
				await writeRowToBigQuery(bigQueryClient, row);
			} catch (bigQueryError) {
				if (bigQueryError instanceof Error) {
					console.error(
						'Failed to insert into BigQuery:',
						bigQueryError.message,
					);
				}
				return { success: false, messageId: record.messageId };
			}

			return { success: true };
		},
	);

	// Wait for all async results to complete
	const results = await Promise.all(asyncResults);

	// Batch return failed items
	// https://docs.aws.amazon.com/lambda/latest/dg/services-sqs-errorhandling.html#services-sqs-batchfailurereporting
	const itemFailures = results
		.filter((result) => !result.success)
		.map((result) => result.messageId);

	return {
		batchItemFailures: itemFailures.map((messageId) => ({
			itemIdentifier: messageId,
		})),
	};
};
