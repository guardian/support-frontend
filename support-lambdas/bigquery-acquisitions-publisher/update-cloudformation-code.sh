#!/bin/sh


cd ../../cdk

yarn synth

# Exit if any of these commands fail, print commands to console
set -ex

aws cloudformation update-stack \
  --capabilities '["CAPABILITY_AUTO_EXPAND", "CAPABILITY_NAMED_IAM", "CAPABILITY_IAM"]'  \
  --stack-name bigquery-acquisitions-publisher-CODE \
  --template-body file://cdk.out/BigqueryAcquisitionsPublisher-CODE.template.json \
  --profile membership

cd -
echo -e "\nStack update has been started, check progress in the AWS console.";
echo -e "https://eu-west-1.console.aws.amazon.com/cloudformation/home";
