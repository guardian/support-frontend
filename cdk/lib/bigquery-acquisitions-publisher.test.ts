import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { BigqueryAcquisitionsPublisher } from "./bigquery-acquisitions-publisher";

describe("The BigqueryAcquisitionsPublisher stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new BigqueryAcquisitionsPublisher(
      app,
      "BigqueryAcquisitionsPublisher-PROD",
      {
        stack: "support",
        stage: "PROD",
        softOptInConsentSetterQueueArn: "arn:aws:sqs:eu-west-1:865473395570:soft-opt-in-consent-setter-queue-PROD",
      }
    );

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
