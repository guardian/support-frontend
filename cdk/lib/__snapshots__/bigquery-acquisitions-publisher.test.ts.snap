// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The BigqueryAcquisitionsPublisher stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
      "GuLambdaErrorPercentageAlarm",
      "GuAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DeadLetterQueueAlarm48D9D867": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":alarms-handler-topic-PROD",
              ],
            ],
          },
        ],
        "AlarmDescription": "There is one or more event in the dead-letters-bigquery-acquisitions-publisher-PROD dead letter queue (DLQ). Check the logs for the error and use the details to confirm that the event was not written to the fact_acquisition_event table in BigQuery. If the event is not in the table then use the DLQ redrive feature to replay the failed event. If the redrive functionality is not used then purge the queue instead or the alarm will remain in an alarm state.
Main queue: https://eu-west-1.console.aws.amazon.com/sqs/v2/home?region=eu-west-1#/queues/https%3A%2F%2Fsqs.eu-west-1.amazonaws.com%2F865473395570%2Fbigquery-acquisitions-publisher-queue-PROD
DLQ: https://eu-west-1.console.aws.amazon.com/sqs/v2/home?region=eu-west-1#/queues/https%3A%2F%2Fsqs.eu-west-1.amazonaws.com%2F865473395570%2Fdead-letters-bigquery-acquisitions-publisher-PROD
Logs: https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Fbigquery-acquisitions-publisher-PROD",
        "AlarmName": "The PROD bigquery-acquisitions-publisher lambda has failed",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "deadlettersbigqueryacquisitionspublisherQueue2B5FA5C1",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Maximum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "EventBusToSQSRule72D08395": {
      "Properties": {
        "Description": "Send all events to SQS",
        "EventBusName": {
          "Ref": "acquisitionsbusPROD768AFDE3",
        },
        "EventPattern": {
          "region": [
            "eu-west-1",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "bigqueryacquisitionspublisherQueue3F783231",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "SoftOptInToSQSRuleA7A6FB25": {
      "Properties": {
        "Description": "Send all events received via support-workers onto soft opt-in SQS queue",
        "EventBusName": {
          "Ref": "acquisitionsbusPROD768AFDE3",
        },
        "EventPattern": {
          "region": [
            "eu-west-1",
          ],
          "source": [
            "support-workers.1",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": "arn:aws:sqs:eu-west-1:865473395570:soft-opt-in-consent-setter-queue-PROD",
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "detail-identityId": "$.detail.identityId",
                "detail-productName": "$.detail.productName",
                "detail-zuoraSubscriptionNumber": "$.detail.zuoraSubscriptionNumber",
              },
              "InputTemplate": "{"subscriptionId":<detail-zuoraSubscriptionNumber>,"identityId":<detail-identityId>,"eventType":"Acquisition","productName":<detail-productName>,"previousProductName":null}",
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "acquisitionsbusPROD768AFDE3": {
      "Properties": {
        "Name": "acquisitions-bus-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::Events::EventBus",
    },
    "acquisitionsbusPRODarchiveArchiveCA0B270E": {
      "Properties": {
        "ArchiveName": "acquisitions-bus-PROD-archive",
        "Description": "Archive for all events sent to acquisitions-bus-PROD-archive",
        "EventPattern": {
          "account": [
            "account",
          ],
          "region": [
            "eu-west-1",
          ],
        },
        "RetentionDays": 90,
        "SourceArn": {
          "Fn::GetAtt": [
            "acquisitionsbusPROD768AFDE3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Events::Archive",
    },
    "bigqueryacquisitionspublisherLambdaD7F059E6": {
      "DependsOn": [
        "bigquerytos3roleDefaultPolicy0D20CB1A",
        "bigquerytos3role8B775EB2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/PROD/bigquery-acquisitions-publisher/index.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "bigquery-acquisitions-publisher",
            "STACK": "support",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "bigquery-acquisitions-publisher-PROD",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "Text",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "bigquerytos3role8B775EB2",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "bigquery-acquisitions-publisher",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "bigqueryacquisitionspublisherLambdaErrorPercentageAlarmForLambdaA9607597": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":alarms-handler-topic-PROD",
              ],
            ],
          },
        ],
        "AlarmDescription": "Check the logs for details https://eu-west-1.console.aws.amazon.com/cloudwatch/home?region=eu-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Fbigquery-acquisitions-publisher-PROD",
        "AlarmName": "big-query-acquisition-publisher lambda has failed",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "bigqueryacquisitionspublisherLambdaD7F059E6",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "bigqueryacquisitionspublisherLambdaD7F059E6",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "bigqueryacquisitionspublisherLambdaD7F059E6",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "bigqueryacquisitionspublisherLambdaSqsEventSourceBigqueryAcquisitionsPublisherPRODbigqueryacquisitionspublisherQueue4510DFFB25270BDD": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "bigqueryacquisitionspublisherQueue3F783231",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "bigqueryacquisitionspublisherLambdaD7F059E6",
        },
        "FunctionResponseTypes": [
          "ReportBatchItemFailures",
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "bigqueryacquisitionspublisherQueue3F783231": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "bigquery-acquisitions-publisher-queue-PROD",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "deadlettersbigqueryacquisitionspublisherQueue2B5FA5C1",
              "Arn",
            ],
          },
          "maxReceiveCount": 1,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VisibilityTimeout": 120,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "bigqueryacquisitionspublisherQueuePolicy28887A75": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "EventBusToSQSRule72D08395",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "bigqueryacquisitionspublisherQueue3F783231",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "bigqueryacquisitionspublisherQueue3F783231",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "bigquerytos3role8B775EB2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "bq-acq-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "bigquerytos3roleDefaultPolicy0D20CB1A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/bigquery-acquisitions-publisher/PROD/gcp-wif-credentials-config",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "bigqueryacquisitionspublisherQueue3F783231",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/PROD/bigquery-acquisitions-publisher/index.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/bigquery-acquisitions-publisher",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/bigquery-acquisitions-publisher/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "bigquerytos3roleDefaultPolicy0D20CB1A",
        "Roles": [
          {
            "Ref": "bigquerytos3role8B775EB2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "deadlettersbigqueryacquisitionspublisherQueue2B5FA5C1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "dead-letters-bigquery-acquisitions-publisher-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-frontend",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
